profiler/internal/cmemprof/profiler.c

```diff
@@ -0,0 +1,158 @@
+	safe_fpunwind = 1;
+}
+
+void profile_allocation(size_t size) {
+	size_t rate = atomic_load_explicit(&sampling_rate, memory_order_relaxed);
+	if (rate == 0) {
+		return;
+	}
+	if (should_sample(rate, size) != 0) {
+		recordAllocationSample(size);
+	}
+}
+
+// is_unsafe_call checks whether the given return address is inside a function
+// from which it is unsafe to call back into Go code.
+//
+// There are two such functions:
+//
+// * x_cgo_thread_start 
+//      This function creates a new OS thread for cgo programs.  Calling back
+//      into Go requires an "m" (OS thread) with a corresponding "g" (goroutine)
+//      to run the Go code. But x_cgo_thread_start creates an "m" and calls
+//      malloc (meaning it might be profiled) before the "m" has a "g". This
+//      causes a crash.
+// * C.malloc
+//      The cgo tool generates wrapper functions for Go programs to call C code.
+//      Since malloc is required by other cgo builtins like C.CString need to
+//      call malloc internally, cgo generates a special hand-written wapper for
+//      malloc. This wrapper lacks checks for whether the calling goroutine's 
+//      stack has grown, which can happen if the C code calls back into Go. This
+//      check is needed so the C function's return value can be returned on the
+//      goroutine's stack. If the stack grows and malloc's return value goes in
+//      the wrong place, the program can crash.
+static int is_unsafe_call(void *ret_addr) {
+	// TODO: Cache this whole check?
+	Dl_info info = {0};
+	if (dladdr(ret_addr, &info) == 0) {
+		return 0;
+	}
+	const char *s = info.dli_sname;
+	cgo_heap_profiler_debug("checking symbol %p: function %s", ret_addr, s);
+	if (s == NULL) {
+		return 0;
+	}
+	if (strcmp(s, "x_cgo_thread_start") == 0) {
+		cgo_heap_profiler_debug("function %s is unsafe", s);
+		return 1;
+	}
+        // Each package which calls C.malloc gets its own malloc wrapper. The
+        // symbol will have a name like "_cgo_<PREFIX>_Cfunc__Cmalloc", where
+        // <PREFIX> is a unique value for each package. We just look for the
+        // suffix.
+	int n = strlen(s);
+	const char *suffix = "_Cfunc__Cmalloc";
+	const int m = strlen(suffix);
+	if (n < m) {
+		return 0;
+	}
+	if (strcmp(s + (n - m), suffix) == 0) {
+		cgo_heap_profiler_debug("function %s is unsafe", s);
+		return 1;
+	}
+	return 0;
+}
+
+// profile_allocation_checked is similar to profile_allocation, but has a
+// fallback for frame pointer unwinding if the provided return address points to
+// a function where profiling by calling into Go is unsafe. (See is_unsafe_call)
+// The return address must be passed in this way because
+// __builtin_return_address(n) may crash for n > 0. This means we can't check
+// the return address anywhere deeper in the call stack and must check in the
+// malloc wrapper.
+void profile_allocation_checked(size_t size, void *ret_addr) {
+	size_t rate = atomic_load_explicit(&sampling_rate, memory_order_relaxed);
```

Comment:
I'll admit that reading [C's memory_order model](https://en.cppreference.com/w/c/atomic/memory_order) gives me headaches, but if I understand it correctly `memory_order_relaxed` has no guarantee to ever observe `return atomic_store(&sampling_rate, hz);` below unless it was called from the same thread?

Reply:
Yeah, I reached for `memory_order_relaxed` mainly because I figured the only thing we need to know in these functions is whether profiling is on or off, and that memory order imposes the fewest restrictions on the program. This function is called for every allocation so I want it to have as little overhead as possible. However, dealing with memory order is really hairy. I'll think hard about what memory order we need to make this work like it should, and add comments to explain why.

Also that `return` does nothing since the return type of that function is `void`... surprised the compiler didn't complain. 🤷

Reply:
Done, opted for the default memory order.